### PR TITLE
Rhythm test

### PR DESCRIPTION
**Pull request by RhythmAgg**
Code Assessment integration

### CODE REVIEW

File: exp_build/experiment.js [hunk 1]
```diff
@@ -192,6 +196,15 @@ class Experiment {
       bp: Config.build_path(this.src) + "/",
     };
 
+    if(options.codeditor) {
```
Comment (raj-vlabs) [inline exp_build/experiment.js:199]: This processing is not expected to happen here. This is supposed to happen in plugin.js in a generic manner.

File: exp_build/plugin.js [hunk 2]
```diff
@@ -99,12 +99,28 @@ class Plugin {
       shell.exec("mkdir plugins");
     }
     pluginConfig.map((plugin) => {
+      if(plugin.label == 'Code Assessment') {
```
Comment (raj-vlabs) [inline exp_build/plugin.js:102]: Please remove this code. If each plugin introduces logic specific to itself in the build framework then this model would not scale. Plugins should be processed using generic code and specific functionality should depend on the configuration parameters.

File: exp_build/task.js [hunk 3]
```diff
@@ -282,6 +282,23 @@ class Task extends Unit {
           }
         }
         break;
+      
+      case ContentTypes.COMPONENT:
+        page_data.isComponent = true;
+        let css_module = [];
+        let js_module = [];
+        exp_info.codeassessment_js_modules.forEach(jsmodule => {
```
Comment (raj-vlabs) [inline exp_build/task.js:290]: Please remove the code-assessment specific code from here and build the component type in a generic manner

**Review by raj-vlabs — CHANGES_REQUESTED** [review]

File: default-experiment-descriptor.json [hunk 4]
```diff
@@ -3,6 +3,11 @@
   "label": "",
   "basedir": ".",
   "LaTeXinMD": false,
+  "code-assessment-options": {
+    "includeCodeAssessment": true,
```
[the remark this answers is not in the document]
Comment (RhythmAgg) [reply default-experiment-descriptor.json:7]: done

File: default-experiment-descriptor.json [hunk 5]
```diff
@@ -3,6 +3,11 @@
   "label": "",
   "basedir": ".",
   "LaTeXinMD": false,
+  "code-assessment": {
```
Comment (raj-vlabs) [inline default-experiment-descriptor.json:6]: default experiment descriptor should have the code editor as false. The developer will turn it to true if they want

**Review by raj-vlabs — CHANGES_REQUESTED** [review]
There are existing review comments which have not been resolved. Please resolve them and resubmit for review

File: default-experiment-descriptor.json [hunk 6]
```diff
@@ -3,6 +3,11 @@
   "label": "",
   "basedir": ".",
   "LaTeXinMD": false,
+  "code-assessment-options": {
```
[the remark this answers is not in the document]
Comment (RhythmAgg) [reply default-experiment-descriptor.json]: resolved

File: default-experiment-descriptor.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (RhythmAgg) [reply default-experiment-descriptor.json]: resolved

File: exp_build/experiment.js (hunk 1 above)
Comment (RhythmAgg) [reply exp_build/experiment.js:209]: resolved

File: exp_build/exp_gen.js [hunk 7]
```diff
@@ -31,12 +31,34 @@ function run(src, lab_data, build_options) {
...
+  if(code_assessment.includeCodeAssessment)
+  {
+    if (!shell.test("-f", Experiment.codeditorPath(src))) {
+      shell.cp(
+        path.resolve(Config.Experiment.default_codeditor),
+        path.resolve(this.src, Experiment.codeditorPath(src))
+      );
+    }
+    log.info("Code Editor included")
+    build_options.codeditor = true;
```
[the remark this answers is not in the document]
Comment (RhythmAgg) [reply exp_build/exp_gen.js:47]: resolved

File: exp_build/plugin.js (hunk 2 above)
Comment (RhythmAgg) [reply exp_build/plugin.js:102]: done

File: exp_build/task.js (hunk 3 above)
Comment (RhythmAgg) [reply exp_build/task.js:290]: done

File: default-experiment-descriptor.json (hunk 5 above)
Comment (RhythmAgg) [reply default-experiment-descriptor.json:6]: done

File: exp_build/experiment.js [hunk 8]
```diff
@@ -246,6 +269,17 @@ class Experiment {
     };
     this.descriptor.units.push(contributors);
   }
+
+  includeCodeEditor(position = 6) {
```
Comment (raj-vlabs) [inline exp_build/experiment.js:273]: Position is 5 in the default experiment descriptor. Why is it defaulted to 6 here? It should be picked from the default experiment descriptor.

File: exp_build/templates/partials/content.handlebars [hunk 9]
```diff
@@ -12,6 +12,9 @@
   {{#if isText}}
     {{{content}}}
   {{/if}}
+  {{#if isComponent}}
```
Comment (raj-vlabs) [inline exp_build/templates/partials/content.handlebars:15]: If we only check for isComponent and include code editor, then we will not be able to add more components in future. Please design it in a future-proof manner

File: validation/schemas/codeAssessment.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
+module.exports = {
```
Comment (raj-vlabs) [inline validation/schemas/codeAssessment.js:1]: File names should not be in camel case. Please rename it as code-assessment.js

File: validation/schemas/codeAssessment.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    $schema: "http://json-schema.org/draft-07/schema#",
+    $id: "https://virtual.labs/schemas/codeAssessment.json",
```
Comment (raj-vlabs) [inline validation/schemas/codeAssessment.js:3]: change the file name to code-assessment.json

**Review by raj-vlabs — CHANGES_REQUESTED** [review]
Please resolve all the old and new comments and resubmit for review.

File: exp_build/experiment.js (hunk 1 above)
Comment (RhythmAgg) [reply exp_build/experiment.js:209]: Sir how to make the processing generic? The plugins with defined scope are handle together whereas these component based plugins have their own build option. Hence we will exclusiveyly have to check for their option and then proceed further

Comment (RhythmAgg) [reply exp_build/experiment.js:209]: I havent made this into generic as of now

**Review by raj-vlabs — APPROVED** [review]

**Pull request merged**